core/src/main/scala/org/http4s/internal/parsing/Rfc3986.scala

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright 2013-2020 http4s.org
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.http4s.internal.parsing
+
+import cats.parse.Parser.{char, charIn, string1}
+import cats.parse.{Parser, Parser1}
+import cats.parse.Rfc5234.{digit, hexdig}
+import org.http4s.Uri.Ipv6Address
+
+/** Common rules defined in Rfc3986
+  *
+  * @see [[https://tools.ietf.org/html/rfc3986]]
+  */
+private[http4s] object Rfc3986 {
+
+  val ipv6: Parser1[Ipv6Address] = {
+
+    def toIpv6(lefts: collection.Seq[Short], rights: collection.Seq[Short]): Ipv6Address =
+      (lefts ++ collection.Seq.fill(8 - lefts.size - rights.size)(0.toShort) ++ rights) match {
+        case collection.Seq(a, b, c, d, e, f, g, h) =>
+          Ipv6Address(a, b, c, d, e, f, g, h)
+      }
+
+    val h16: Parser1[Short] = {
+      (hexdig ~ hexdig.? ~ hexdig.? ~ hexdig.?).string.map { (s: String) => java.lang.Integer.parseInt(s, 16).toShort }
+    }
+
+    val decOctet = (char('1') ~ digit ~ digit)
+      .orElse1(char('2') ~ charIn('0' to '4') ~ digit)
+      .orElse1(string1("25") ~ charIn('0' to '5'))
+      .orElse1(charIn('1' to '9') ~ digit)
+      .orElse1(digit).string.map(_.toInt.toByte)
+
+    val ls32: Parser1[(Short, Short)] = {
+      val option1 = (h16 ~ char(':').void ~ h16).map { case ((l, _), r) => (l, r) }
+      val dot = char('.')
+      val option2: Parser1[(Short, Short)] = (decOctet ~ dot ~ decOctet ~ dot ~ decOctet ~ dot ~ decOctet).map {
+        case ((((((a, _), b), _), c), _), d) =>
+          ((a << 8) | b).toShort -> ((c << 8) | d).toShort
+      }
+      option1.orElse1(option2)
+    }
+
+    val colon = char(':')
```

Comment:
You're using the `char` expression elsewhere.  Might as well take advantage of your constant.